routing/osrm_path_segment_factory.cpp

```diff
@@ -143,40 +144,77 @@ void OsrmPathSegmentFactory(RoutingMapping & mapping, Index const & index, RawPa
-  if (isStartNode && isEndNode)
+  size_t const startIndex = isStartNode ? findIntersectingSeg(startGraphNode.segment) : 0;
+  size_t const endIndex = isEndNode ? findIntersectingSeg(endGraphNode.segment) + 1 : buffer.size();
+  // @TODO(bykoianko) When start and end points belong to the same NodeID it's possible
+  // to get endIndex <= startIndex. Production version returns route building error.
+  // Debug version crashes with an assert. The test to show it USALosAnglesAriaTwentyninePalmsHighwayTimeTest.
+  LoadPathGeometry(buffer, startIndex, endIndex, index, mapping, startGraphNode, endGraphNode, isStartNode,
+                   isEndNode, loadedPathSegment);
+
+  LoadedPathSegment fullPathSegment;
+  LoadPathGeometry(buffer, 0, buffer.size(), index, mapping, FeatureGraphNode(), FeatureGraphNode(),
+                   false /* isStartNode */, false /* isEndNode */, fullPathSegment);
+  double const fullPathLengthM = PathLengthM(fullPathSegment.m_path);
+  if (fullPathLengthM == 0.0 || loadedPathSegment.m_path.empty())
+  {
+    loadedPathSegment.m_weight = 0.0;
+    return;
+  }
+
+  // Full weight (time needed to path NodeID).
+  double fullWeight = 0; /* |fullWeight * kOSRMWeightToSecondsMultiplier| is time in seconds. */
+  if (isStartNode)
+  {
+    // Note. It's possible that isStartNode == true and isEndNode == true.
+    // In that case it's ok to follow this branch of if because in that case
+    // startGraphNode.node.forward_node_id == endGraphNode.node.forward_node_id.
+    ASSERT(isStartNode != isEndNode
+           || startGraphNode.node.forward_node_id == endGraphNode.node.forward_node_id
+           || startGraphNode.node.forward_node_id == endGraphNode.node.reverse_node_id, ());
+    fullWeight = (osrmPathSegment.node == startGraphNode.node.forward_node_id)
+        ? startGraphNode.node.forward_offset
+        : startGraphNode.node.reverse_offset;
+  }
+  else /* isEndNode */
   {
-    double const forwardWeight = (osrmPathSegment.node == startGraphNode.node.forward_node_id)
-                                     ? startGraphNode.node.forward_weight
-                                     : startGraphNode.node.reverse_weight;
-    double const backwardWeight = (osrmPathSegment.node == endGraphNode.node.forward_node_id)
-                                      ? endGraphNode.node.forward_weight
-                                      : endGraphNode.node.reverse_weight;
-    double const wholeWeight = (osrmPathSegment.node == startGraphNode.node.forward_node_id)
-                                   ? startGraphNode.node.forward_offset
-                                   : startGraphNode.node.reverse_offset;
-    // Sum because weights in forward/backward_weight fields are negative. Look osrm_helpers for
-    // more info.
-    loadedPathSegment.m_weight = wholeWeight + forwardWeight + backwardWeight;
+    fullWeight = (osrmPathSegment.node == endGraphNode.node.forward_node_id)
```

Comment:
Ok. I see. But I think, that it will be better to repair an unpredictable state for two reasons:
1. Because you need to calculate the partPathLengthM here for the second time (first was on forwardWeight/reverseWeight calculation).
2. Because offset-weight calculation has impact on first routing edges (it can takes the wrong route).

You can see all weight calculations in the routing/osrm_helpers.cpp file. Function: CalculateWeight
Forward and reverse weights were calculated based on intersection of the navigation route and the NodeID here.